packages/api/src/utils/externalApi.ts

```diff
@@ -0,0 +1,190 @@
+      aggFn,
+      where: where ?? '',
+      groupBy: groupBy ?? [],
+      ...(field ? { field } : {}),
+      ...(numberFormat ? { numberFormat } : {}),
+      ...(metricDataType ? { metricDataType } : {}),
+    };
+
+    return series;
+  } else if (type === 'table') {
+    if (aggFn == null) {
+      throw new Error('aggFn must be set for table chart');
+    }
+
+    const series: z.infer<typeof tableChartSeriesSchema> = {
+      type: 'table',
+      table,
+      aggFn,
+      where: where ?? '',
+      groupBy: groupBy ?? [],
+      sortOrder: sortOrder ?? 'desc',
+      ...(field ? { field } : {}),
+      ...(numberFormat ? { numberFormat } : {}),
+      ...(metricDataType ? { metricDataType } : {}),
+    };
+
+    return series;
+  } else if (type === 'number') {
+    if (aggFn == null) {
+      throw new Error('aggFn must be set for number chart');
+    }
+
+    const series: z.infer<typeof numberChartSeriesSchema> = {
+      type: 'number',
+      table,
+      aggFn,
+      where: where ?? '',
+      ...(field ? { field } : {}),
+      ...(numberFormat ? { numberFormat } : {}),
+      ...(metricDataType ? { metricDataType } : {}),
+    };
+
+    return series;
+  } else if (type === 'histogram') {
+    const series: z.infer<typeof histogramChartSeriesSchema> = {
+      type: 'histogram',
+      table,
+      where: where ?? '',
+      ...(field ? { field } : {}),
+      ...(metricDataType ? { metricDataType } : {}),
+    };
+
+    return series;
+  } else if (type === 'search') {
+    const series: z.infer<typeof searchChartSeriesSchema> = {
+      type: 'search',
+      fields: fields ?? [],
+      where: where ?? '',
+    };
+
+    return series;
+  } else if (type === 'markdown') {
+    const series: z.infer<typeof markdownChartSeriesSchema> = {
+      type: 'markdown',
+      content: content ?? '',
+    };
+
+    return series;
+  }
+
+  throw new Error(`Invalid chart type ${type}`);
+};
+
+export const translateExternalChartToInternalChart = (
+  chartInput: z.infer<typeof externalChartSchemaWithId>,
+): z.infer<typeof chartSchema> => {
+  const { id, x, name, y, w, h, series, asRatio } = chartInput;
+  return {
+    id,
+    name,
+    x,
+    y,
+    w,
+    h,
+    seriesReturnType: asRatio ? 'ratio' : 'column',
+    series: series.map(s => translateExternalSeriesToInternalSeries(s)),
+  };
+};
+
+const translateChartDocumentToExternalChart = (
+  chart: z.infer<typeof chartSchema>,
+): z.infer<typeof externalChartSchemaWithId> => {
+  const { id, x, name, y, w, h, series, seriesReturnType } = chart;
+  return {
+    id,
+    name,
+    x,
+    y,
+    w,
+    h,
+    asRatio: seriesReturnType === 'ratio',
+    series: series.map(s => {
+      const {
+        type,
+        table,
+        aggFn,
+        field,
+        where,
+        groupBy,
+        sortOrder,
+        content,
+        numberFormat,
+      } = s;
+
+      return {
+        type,
+        data_source: table === 'metrics' ? 'metrics' : 'events',
```

Comment:
do we want to make this guy camelcase ?

Reply:
🤦 yeah good point... there's a few places where I introduced the snake case. Let me do it in a follow up PR after this and the TF one is merged so I can change it everywhere at once.